pydatastructs/miscellaneous_data_structures/queue.py

```diff
@@ -364,3 +367,174 @@ def peek(self):
     @property
     def is_empty(self):
         return self.items.is_empty
+
+class ArrayDeque(Queue, DynamicOneDimensionalArray):
+    """
+    Represents Deque datastracture implemented using Array
+
+    Parameters
+    ==========
+
+    dtype: type
+        A valid object type.
+    size: int
+        The number of elements in the array.
+    elements: list/tuple
+        The elements in the array, all should
+        be of same type.
+    init: a python type
+        The inital value with which the element has
+        to be initialized. By default none, used only
+        when the data is not given.
+    load_factor: float, by default 0.25
+        The number below which if the ratio, Num(T)/Size(T)
+        falls then the array is contracted such that at
+        most only half the positions are filled.
+
+    Raises
+    ======
+
+    ValueError
+        When the number of elements in the list do not
+        match with the size.
+        More than three parameters are passed as arguments.
+        Types of arguments is not as mentioned in the docstring.
+        The load factor is not of floating point type.
+
+    IndexError
+        When pop/popleft is used on an empty Deque
+
+    Note
+    ====
+
+    At least one parameter should be passed as an argument along
+    with the dtype.
+    Num(T) means the number of positions which are not None in the
+    array.
+    Size(T) means the maximum number of elements that the array can hold.
+
+    Examples
+    ========
+
+    >>> from pydatastructs import Queue
+    >>> q = Queue(implementation='deque', dtype=int, items=[1,2,3])
+    >>> q.pop()
+    3
+    >>> q.popleft()
+    1
+    >>> q.append(1)
+    >>> q.appendleft(3)
+    >>> [q[i] for i in range(len(q))]
+    [3, 2, 1]
+    """
+
+    __slots__ = ['_load_factor', '_num', '_last_pos_filled', '_size', '_rear', '_front', '_num']
+
+    def __new__(cls, dtype, *args, **kwargs):
+        obj = DynamicOneDimensionalArray.__new__(cls, dtype, *args, **kwargs)
+        obj._front = obj._last_pos_filled
+        obj._rear = 0
+        obj._num = obj._front + 1
+        return obj
+
+    @property
+    def rear(self):
+        return (len(self._data) + self._rear) % len(self._data)
+
+    @property
+    def front(self):
+        return (len(self._data) + self._front) % len(self._data)
+
+    @front.setter
+    def front(self, val):
+        self._front = val
+
+    @rear.setter
+    def rear(self, val):
+        self._rear = val
+
+    @property
+    def size(self):
+        return self._num
+
+    @property
+    def is_empty(self):
+        return self.size == 0
+
+    @property
+    def is_full(self):
+        return self.size == len(self._data)
+
+    def __getitem__(self, idx):
+        idx = (self.rear + idx) % len(self._data)
+        return self._data[idx]
+
+    def __setitem__(self, idx, val):
+        idx = (self.rear + idx) % len(self._data)
+        self._data[idx] = val
```

Comment:
In queue, elements are accessed from end points so these methods shouldn't be defined.

Reply:
helps a lot in rest of the implementation since I dont have to do modular arithmetic every time I push or pop

Reply:
At least they shouldn't be public. You can define helper methods, though 2 line is not much for code indirection.

Reply:
and these functions are very helpful if u want to change the value of some key in the queue, the user does not have to think about rear and front they can just start from 0 and go up to self.size. It actually makes the code a lot cleaner.

Reply:
> the user does not have to think about rear and front

If someone wants to change the value of a key, they could have used some other data structure.

Reply:
if we dont override it here then if user uses the index it will use DODA's __getitem__ and __setitem__ which will break the logic, and anyway it isnt breaking the already set datastructures and it helps a lot in development also as you can see in append/ appendleft. It makes the code readable, it helps to remove unnecessary bugs and there are only 2 small methods in question.

Reply:
and sir it also helps other developers to understand the code.